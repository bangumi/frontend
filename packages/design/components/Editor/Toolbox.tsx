import React, { FC } from 'react'
import { Bold, Image, Italic, Link, Size, Underscore } from '@bangumi/icons'

const Toolbox: FC = () => {
  return (
    <div className="bgm-editor__header">
      <Bold />
      <Italic />
      <Underscore />
      <Image />
      <Link />
      <Size />
    </div>
  )
}

export default Toolbox
