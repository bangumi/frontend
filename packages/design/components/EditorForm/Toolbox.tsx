import type { FC } from 'react';
import React, { memo } from 'react';

import { Bold, Image, Italic, Link, Size, Underscore } from '@bangumi/icons';

interface ToolboxProps {
  handleClickEvent: (type: string, payload?: unknown) => void;
  style?: React.CSSProperties;
}

const Toolbox: FC<ToolboxProps> = ({ handleClickEvent, style }) => {
  return (
    <div className='bgm-editor__toolbox' style={style}>
      <Bold
        data-testid='bold'
        onClick={() => {
          handleClickEvent('bold');
        }}
        title='粗体 [Ctrl+B]'
      />
      <Italic
        data-testid='italic'
        onClick={() => {
          handleClickEvent('italic');
        }}
        title='斜体 [Ctrl+I]'
      />
      <Underscore
        data-testid='underscore'
        onClick={() => {
          handleClickEvent('underscore');
        }}
        title='下划线 [Ctrl+U]'
      />
      <Image
        data-testid='image'
        onClick={() => {
          handleClickEvent('image');
        }}
        title='图片 [Ctrl+P]'
      />
      <Link
        data-testid='link'
        onClick={() => {
          handleClickEvent('link');
        }}
        title='链接 [Ctrl+L]'
      />
      <Size
        data-testid='size'
        onClick={() => {
          handleClickEvent('size');
        }}
        title='文字尺寸 [Ctrl+S]'
      />
    </div>
  );
};

export default memo(Toolbox);
