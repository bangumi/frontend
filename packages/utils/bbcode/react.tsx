import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
