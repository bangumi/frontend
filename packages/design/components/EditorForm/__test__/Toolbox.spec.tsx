import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';

import Toolbox from '../Toolbox';

describe('EditorForm > Toolbox', () => {
  it('render with default classNames', () => {
    const { asFragment } = render(<Toolbox handleClickEvent={null!} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Toolbox style props', () => {
    const { container } = render(<Toolbox handleClickEvent={null!} style={{ display: 'none' }} />);
    expect(container.firstChild).toHaveStyle('display: none');
  });

  it('Toolbox handleClickEvent props', () => {
    const handleClickEvent = vi.fn();
    render(<Toolbox handleClickEvent={handleClickEvent} />);
    for (const type of ['bold', 'italic', 'underscore', 'image', 'link', 'size']) {
      fireEvent.click(screen.getByTestId(type));
      expect(handleClickEvent).toHaveBeenCalledWith(type);
    }
  });
});
