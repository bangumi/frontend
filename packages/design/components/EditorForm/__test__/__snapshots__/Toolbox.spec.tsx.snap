// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorForm > Toolbox render with default classNames 1`] = `
<DocumentFragment>
  <div
    class="bgm-editor__toolbox"
  >
    <div
      data-testid="bold"
    />
    <div
      data-testid="italic"
    />
    <div
      data-testid="underscore"
    />
    <div
      data-testid="image"
    />
    <div
      data-testid="link"
    />
    <div
      data-testid="size"
    />
  </div>
</DocumentFragment>
`;
